rust/tests/vacuum_test.rs

```diff
@@ -76,3 +85,261 @@ async fn vacuum_delta_8_0_table() {
 
     assert_eq!(result.files_deleted, empty);
 }
+
+#[tokio::test]
+// Validate vacuum works on a non-partitioned table
+async fn test_non_partitioned_table() {
+    let mut context = TestContext::from_env().await;
```

Comment:
Your assumption on testing one backend at a time is correct. I'd like to hear additional comment for others on which design approach we should move forward with. 

Here are my thoughts on the structure of the object store tests. It looks like the generic tests run in series. This can be resolved by wrapping each generic test for the specific backend but requires (num_backends * generic_tests) definitions compared to just (generic_tests) for this approach.

Also, with the current approach, new backends only need to implement a setup and clean up functions and they get all existing and future tests for free without implementing additional code. Compared with object store where multiple call sites need to be updated to incorporate the new generic function or a generic function might be missed when implementing a new backend.

Reply:
> Your assumption on testing one backend at a time is correct.

Ideally, I personally would like to have all backends integration tested in CI. Even more, it would be nice to easily test all backends by simply running:

```bash
docker compose up
cargo test --all-features
```

> This can be resolved by wrapping each generic test for the specific backend but requires (num_backends * generic_tests) definitions compared to just (generic_tests) for this approach.

Ah agreed that it's not ideal that they run in series. In a different project, my approach to that problem was to write macros: https://github.com/wjones127/arrow-datafusion/blob/66e7da07be154986b9416aab1b038ead85533c23/data-access/src/object_store/mod.rs#L263-L293


> Compared with object store where multiple call sites need to be updated to incorporate the new generic function or a generic function might be missed when implementing a new backend.

It might be missing accidentally, but sometimes it can be a feature to have granular control over which test run for which backend, such as when there are features left TODO. So there are benefits and drawbacks.

Reply:
Yeah the only satisfactory way to effectively solve this would involve macros. 
My proposal is to introduce a new attribute-like macro with name `deltalake::test` that would create a wrapper for each backend. To support granularity, specific backends can be specified so only wrappers are created for those backends or an ignore option can specify which backends to omit.

For example
```rust
#[deltalake::test]
pub async fn test_partitioned_table(context: TestContext) { ... }
```
Would expand to
```rust
pub async fn test_partitioned_table(context: TestContext) { ... }

#[cfg(feature = "s3")]
#[tokio::test]
pub async fn test_partitioned_table_s3(context: TestContext) { 
    let context = create_context("s3").await.unwrap();
    test_partitioned_table(context).await;
 }

#[cfg(feature = "azure")]
#[tokio::test]
pub async fn test_partitioned_table_azure(context: TestContext) { 
    let context = create_context("azure").await.unwrap();
    test_partitioned_table(context).await;
 }
```
Then if we want to limit the scope to specific backend the following can be specified
```rust
#[deltalake::test(backend=["s3", "gcs"])
pub async fn test1(context: TestContext) { ... }
```

Let me know if you think this approach is sensible and worth pursuing.

Reply:
I like that idea. It does seem expansive enough that we should pursue it in a follow-up issue, rather than in this PR. Also since we can't use Azurite currently, it's not quite as valuable to have yet. But it does seem like a great solution!